cupy/_core/_gufuncs.py

```diff
@@ -141,6 +142,167 @@ def _validate_normalize_axes(
     return input_axes, output_axes
 
 
+class _OpsRegister:
+    '''
+    Holds the ops for each dtypes signature like ('ff->f', func1)
+    and allows to do look ups for these
+    '''
+    class _Op:
+        def __init__(self, in_types, out_types, func):
+            self.func = func
+            self.in_types = tuple(numpy.dtype(i) for i in in_types)
+            self.out_types = tuple(numpy.dtype(o) for o in out_types)
+            self.sig_str = (''.join(
+                in_t.char for in_t in self.in_types) + '->' + ''.join(
+                    out_t.char for out_t in self.out_types))
+
+    def __init__(self, signatures, default_func, nin, nout):
+        self._default_func = default_func
+        self._nin = nin
+        self._nout = nout
+        self. _ops = self._process_signatures(signatures)
+
+    def _sig_str_to_tuple(self, sig):
+        sig = sig.replace(' ', '')
+        toks = sig.split('->')
+        if sig.startswith('->'):
+            ins = ''
+            outs = toks
+        elif sig.endswith('->'):
+            ins = toks
+            outs = ''
+        elif len(toks) != 2:
+            raise ValueError(f'signature {sig} for dtypes is invalid')
+        else:
+            ins, outs = toks
+        return ins, outs
+
+    def _process_signatures(self, signatures):
+        ops = []
+        for sig in signatures:
+            if isinstance(sig, tuple):
+                sig, op = sig
+            else:
+                op = self._default_func
+            ins, outs = self._sig_str_to_tuple(sig)
+            # Check the number of inputs and outputs matches the gufunc sig
+            if len(ins) != self._nin:
+                raise ValueError(
+                    f'signature {sig} for dtypes is invalid number of inputs '
+                    'is not consistent with general signature')
+            if len(outs) != self._nout:
+                raise ValueError(
+                    f'signature {sig} for dtypes is invalid number of inputs '
+                    'is not consistent with general signature')
+
+            ops.append(_OpsRegister._Op(ins, outs, op))
+        return ops
+
+    def _can_cast(self, d1, d2, casting):
```

Comment:
In what case does this function behave differently from `numpy.can_cast` itself?

Reply:
Its the same but with a shortcut, this was taken from kernel.pyx iirc